File: php/src/IdkollenClient.php
<?php

declare(strict_types=1);

namespace Idkollen\Client;

use Idkollen\Client\Endpoints\BankIdNoEndpoint;
use Idkollen\Client\Endpoints\BankIdSeEndpoint;
use Idkollen\Client\Endpoints\DocumentEndpoint;
use Idkollen\Client\Endpoints\FrejaEndpoint;
use Idkollen\Client\Endpoints\FtnEndpoint;
use Idkollen\Client\Endpoints\MitIdEndpoint;
use Idkollen\Client\Endpoints\VippsEndpoint;

final class IdkollenClient
{
    public function __construct(
        private readonly Transport $transport,
    ) {}

    public function bankIdSe(): BankIdSeEndpoint
    {
        return new BankIdSeEndpoint($this->transport);
    }

    public function bankIdNo(): BankIdNoEndpoint
    {
        return new BankIdNoEndpoint($this->transport);
    }

    public function freja(): FrejaEndpoint
    {
        return new FrejaEndpoint($this->transport);
    }

    public function mitId(): MitIdEndpoint
    {
        return new MitIdEndpoint($this->transport);
    }

    public function ftn(): FtnEndpoint
    {
        return new FtnEndpoint($this->transport);
    }

    public function vipps(): VippsEndpoint
    {
        return new VippsEndpoint($this->transport);
    }

    public function document(): DocumentEndpoint
    {
        return new DocumentEndpoint($this->transport);
    }
}

File: swift/Sources/IdkollenClient/IdkollenClient.swift
import Foundation

public final class IdkollenClient {
    public let bankIdSe: BankIdSeEndpoint
    public let bankIdNo: BankIdNoEndpoint
    public let freja: FrejaEndpoint
    public let mitId: MitIdEndpoint
    public let ftn: FtnEndpoint
    public let vipps: VippsEndpoint
    public let document: DocumentEndpoint

    internal init(transport: Transport) {
        self.bankIdSe = BankIdSeEndpoint(transport)
        self.bankIdNo = BankIdNoEndpoint(transport)
        self.freja = FrejaEndpoint(transport)
        self.mitId = MitIdEndpoint(transport)
        self.ftn = FtnEndpoint(transport)
        self.vipps = VippsEndpoint(transport)
        self.document = DocumentEndpoint(transport)
    }
}

public final class IdkollenClientBuilder {
    private let clientId: String
    private let clientSecret: String
    private var env: Environment = .production
    private var customBaseUrl: String?
    private var customSession: URLSession?

    public init(clientId: String, clientSecret: String) {
        self.clientId = clientId
        self.clientSecret = clientSecret
    }

    @discardableResult
    public func environment(_ env: Environment) -> Self {
        self.env = env
        return self
    }

    @discardableResult
    public func baseUrl(_ url: String) -> Self {
        self.customBaseUrl = url
        return self
    }

    @discardableResult
    public func urlSession(_ session: URLSession) -> Self {
        self.customSession = session
        return self
    }

    public func build() -> IdkollenClient {
        let session = customSession ?? URLSession.shared
        let baseUrl = customBaseUrl ?? env.baseUrl
        let transport = Transport(session: session, baseUrl: baseUrl, clientId: clientId, clientSecret: clientSecret)
        return IdkollenClient(transport: transport)
    }
}
